src/cpp/fastdds/domain/DomainParticipantImpl.hpp

```diff
@@ -561,6 +585,53 @@ class DomainParticipantImpl
 
     class MyRTPSParticipantListener : public fastrtps::rtps::RTPSParticipantListener
     {
+        struct Sentry
+        {
+            Sentry(
+                    MyRTPSParticipantListener* listener)
+                : listener_(listener)
+                , on_guard_(false)
+            {
+                if (listener_ != nullptr && listener_->participant_ != nullptr)
+                {
+                    std::lock_guard<std::mutex> _(listener_->participant_->mtx_gs_);
+                    if (listener_->callback_counter_ >= 0)
+                    {
+                        ++listener_->callback_counter_;
+                        on_guard_ = true;
+                    }
+                }
+            }
+
+            ~Sentry()
+            {
+                if (on_guard_ && listener_ != nullptr && listener_->participant_ != nullptr)
```

Comment:
I would leave `if (on_guard_)` alone and assert the value of `listener_->participant_ != nullptr` this way if we introduce inconsistencies they can be readily detected.